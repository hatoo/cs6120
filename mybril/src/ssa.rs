use std::{
    collections::{HashMap, HashSet},
    iter::Successors,
};

use crate::{basic_block::BasicBlock, Argument, Function};

pub struct CfgEntry {
    basic_block: BasicBlock,
    predesessors: HashSet<String>,
    successors: HashSet<String>,
}

pub struct Cfg {
    arguments: Vec<Argument>,
    entry: String,
    graph: HashMap<String, CfgEntry>,
}

impl Cfg {
    pub fn new(function: &Function) -> Self {
        let basic_blocks = BasicBlock::new_blocks(&function.instrs);

        let arguments = function.args.clone().unwrap_or_default();
        let entry = basic_blocks[0][0].label.clone().unwrap();

        let mut predesessors = HashMap::new();
        let mut successors = HashMap::new();

        for block in &basic_blocks {
            let label = block[0].label.clone().unwrap();
            if let Some(nexts) = block.last().unwrap().labels.as_ref() {
                for next in nexts {
                    let next = next.clone();
                    successors
                        .entry(label.clone())
                        .or_insert_with(HashSet::new)
                        .insert(next.clone());
                    predesessors
                        .entry(next)
                        .or_insert_with(HashSet::new)
                        .insert(label.clone());
                }
            }
        }

        let graph = basic_blocks
            .into_iter()
            .map(|block| {
                let label = block[0].label.clone().unwrap();
                let predesessors = predesessors.remove(&label).unwrap_or_default();
                let successors = successors.remove(&label).unwrap_or_default();
                (
                    label,
                    CfgEntry {
                        basic_block: block,
                        predesessors,
                        successors,
                    },
                )
            })
            .collect();

        Self {
            arguments,
            entry,
            graph,
        }
    }

    pub fn reverse_post_order(&self) -> Vec<&str> {
        let mut visited = HashSet::new();
        let mut order = vec![];

        fn rec<'a>(
            label: &'a str,
            visited: &mut HashSet<&'a str>,
            order: &mut Vec<&'a str>,
            graph: &'a HashMap<String, CfgEntry>,
        ) {
            if visited.contains(label) {
                return;
            }
            visited.insert(label);
            for next in &graph[label].successors {
                rec(next, visited, order, graph);
            }
            order.push(label);
        }

        rec(&self.entry, &mut visited, &mut order, &self.graph);
        order.reverse();
        order
    }
}
