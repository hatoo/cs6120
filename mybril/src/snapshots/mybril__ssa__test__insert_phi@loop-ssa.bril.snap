---
source: mybril/src/ssa.rs
expression: "format!(\"{}\\n\\n{} -> {}\\n\\n{}\", txt, orig.1, after.1,\n    bril2txt(json_after.as_str()))"
input_file: mybril/tests/examples/ssa/loop-ssa.bril
---
@main {
.entry:
    i.1: int = const 1;
    jmp .loop;
.loop:
    i.2: int = phi .entry i.1 .body i.3;
    max: int = const 10;
    cond: bool = lt i.2 max;
    br cond .body .exit;
.body:
    i.3: int = add i.2 i.2;
    jmp .loop;
.exit:
    print i.2;
}


31 -> 42

@main {
.entry:
  i.1: int = const 1;
  jmp .loop;
.loop:
  i.1 = phi i.1 .entry;
  i.3 = phi i.3 .body;
  i.2: int = phi i.1 i.3 .entry .body;
  max: int = const 10;
  cond: bool = lt i.2 max;
  br cond .body .exit;
.body:
  i.3: int = add i.2 i.2;
  jmp .loop;
.exit:
  print i.2;
  ret;
}

